### PR TITLE
Added support for Osmdroid in the F-Droid version of RunnerUp

### PR DESCRIPTION
**Pull request by Upabjojr**
The `MapView` widget will now be visible in the free version of RunnerUp. It will link to an Osmdroid widget which is compatible with F-Droid.

This edit allows to view the track on a map with the free version on F-Droid.

The play version will remain identical to the current one.

### CODE REVIEW

File: app/src/main/org/runnerup/view/DetailActivity.java [hunk 1]
```diff
@@ -127,13 +127,10 @@ public class DetailActivity extends AppCompatActivity implements Constants {
     @Override
     public void onCreate(Bundle savedInstanceState) {
         super.onCreate(savedInstanceState);
-        if (BuildConfig.MAPBOX_ENABLED > 0) {
```
Comment (gerhardol) [inline app/src/main/org/runnerup/view/DetailActivity.java:130]: This is still needed for Play version without MapBox key.

**Review by gerhardol — COMMENTED** [review]
Will likely take me some days to get into this...

File: app/src/main/org/runnerup/view/DetailActivity.java (hunk 1 above)
Comment (Upabjojr) [reply app/src/main/org/runnerup/view/DetailActivity.java:130]: What about this:
```java
if (BuildConfig.APPLICATION_ID.contains(".free") || BuildConfig.MAPBOX_ENABLED > 0) {
    ...
}
```
Is there a better way to understand if the code was compiled using the free version?

**Comment by gerhardol** [comment]
Seem to be working fine.
I got some exception at first use but before I could investigate the logs were gone.
The markers could be improved.
Also, if no points at all the zoom level is max, compared to min with MapBox.
This could well be replacing the MapBox inplementation also in Play. In addition to the free aspect, the map usage is limited in MapBox.

**Comment by Upabjojr** [comment]
> The markers could be improved.

Osmdroid appears to have a very limited choice of icons for markers. External icons may be needed.

> This could well be replacing the MapBox inplementation also in Play. In addition to the free aspect, the map usage is limited in MapBox.

Bear in mind that this would use OpenStreetMap's servers in the Play version. The servers are run by donations, while MapBox is a for-profit company. Not sure we should abuse OpenStreetMap's servers while there are alternatives.

**Comment by knarrff** [comment]
> Not sure we should abuse OpenStreetMap's servers while there are alternatives.

I doubt usage of OSM in RunnerUp would even show up in their overall usage. Also, why would it be abuse to use OSM here - wasn't is built to be used?

**Comment by Upabjojr** [comment]
So do you suggest to switch to Osmdroid by default and drop MapBox?

**Comment by knarrff** [comment]
This is not up to me: I am not even a developer. I just found the reasoning a little strange in "Not sure we should abuse OpenStreetMap's servers while there are alternatives."

**Comment by gerhardol** [comment]
The primarily reason for RunnerUp is that is an open source application. That is the reason why I maintain it - I am not really a user myself but I want to keep alternatives open.
(I recently broke my new phone when bring it for a run).
It is occasionally fun to develop, even if I have got turned off by all changes Google enforces.

So just using free SW is in line with my intention. I am am not strict about it, there are hidden upload keys too. The MapBox usage is not very high. 5K users in Play, the free MapBox account has currently 634/25000 monthly users.
Usage in F-Droid is unknown, I have no statistics at all there.

For now we can keep MapBox in Play, may drop it at next API change they make.

**Comment by Upabjojr** [comment]
> Usage in F-Droid is unknown, I have no statistics at all there.

I guess there are probably less users on F-Droid than on Play. I don't have Play on my phone so I use the F-Droid version, but I image most people do install the app from Play.

File: app/src/free/org/runnerup/util/MapWrapper.java [hunk 2]
```diff
@@ -17,35 +17,135 @@
...
 import org.runnerup.common.util.Constants;
-import org.runnerup.util.Formatter;
+import org.runnerup.db.entities.LocationEntity;
 
+import java.util.LinkedList;
+import java.util.List;
 
 public class MapWrapper implements Constants {
 
+    public static final boolean USING_OSMDROID = true;
```
Comment (gerhardol) [inline app/src/free/org/runnerup/util/MapWrapper.java:43]: Use buildConfigField instead to set this.

File: app/src/free/org/runnerup/util/MapWrapper.java [hunk 3]
```diff
@@ -17,35 +17,135 @@
...
+                markerEnd.setTextIcon(context.getString(R.string.Finished));
+
+                markerStart.setAnchor(Marker.ANCHOR_CENTER, Marker.ANCHOR_BOTTOM);
+                markerEnd.setAnchor(Marker.ANCHOR_CENTER, Marker.ANCHOR_BOTTOM);
+
+                mapView.getOverlays().add(markerStart);
+                mapView.getOverlays().add(markerEnd);
+            }
+
+            mapView.getOverlays().add(route);
```
Comment (gerhardol) [inline app/src/free/org/runnerup/util/MapWrapper.java:132]: Should this be added only if there are points?

**Review by gerhardol — COMMENTED** [review]
Just a few comments

**Comment by Upabjojr** [comment]
I think this feature should be added to the F-Droid version at first, as I'm not sure it's as complete as the MapBox widget. This can still be changed in the future.

**Review by gerhardol — APPROVED** [review]

**Pull request merged**